### ISSUE TITLE
[New build - MECH PREVIEW] 24w24a: Mech Round 2, Hotfixes

### ISSUE THREAD

**Issue by tsunamayo** [tsunamayo/Starship-EVO#5803]
New improvements to the mechs:

- Custom tag: can input a custom limb type for custom limb chain
- Feet alignment: feet will align to the terrain allowing walking on slope
- SFX and VFX added.
- Animation template added. For now there is now built-in template but we will work on that!
I will try to work on a retargeting system so various animation are still compatible with different rotator orientations.
- View terrain in preview windows
- Replay works both way.
- Height and speed graph. Ideal to debug your walking cycle.
- Feet depenetration improved
- Launch to orbit bug fixed.

Hotfixes build.

Hotfixes:
#5789 Docking port: Is Docked event not working properly.
#5788 Tri girder minimum grid size.
#5774 Docking port: cannot transfert items
#5775 Gravity don't apply after undocking
#5777 Power cell issues, collision and on delete
#5786 Life support dont refill in atmosphere.
#5798 Symmetry on assembler is incorrect

**Comment by ProPeach** [comment]
Happy to confirm the launch to orbit is fixed! Even on mechs where Cancel Slip Amount didn't fix it too.

Looks like this update also includes symmetry for the leg animations which is awesome, I don't see it in these patch notes though.

Great update all round, looking forward to testing further

**Comment by tsunamayo** [comment]
@ProPeach yes I forgot about symmetry.

**Comment by ProPeach** [comment]
Yes I noticed that! Very very cool, I was surprised to see  that feature in so quickly as it sounded like it would be a pain

**Comment by BigBadKangaroo** [comment]
The mechs feel a lot better now. Still a bit clunky but much better than before.

It seems like the automated copy&paste of the joint rotations is a bit bugged aswell.
The left side is set by me and works actually really well.
The right side is copied and the foot has a 0 vaule and it "snaps" when the cicle starts a new rotation:

https://github.com/tsunamayo/Starship-EVO/assets/81495956/32aaac81-0642-4d88-a1f6-3d2f303efea4


https://github.com/tsunamayo/Starship-EVO/assets/81495956/f530fc48-505a-4497-a0de-da67c302c3c4

That's the mech: 
[Mech_5.zip](https://github.com/user-attachments/files/15841059/Mech_5.zip)

**Comment by asanagisae** [comment]
The new build certainly feels better.

https://github.com/tsunamayo/Starship-EVO/assets/10069469/6c3b53e1-d9fe-4850-b2e3-e78a167738d9

The mechawheel never appears in the mech computer preview:

![image](https://github.com/tsunamayo/Starship-EVO/assets/10069469/54103440-ce0f-4577-8a0b-f420ed8a5668)

If I try to assign it an animation template, it deletes all of my key frame data from idle and walk.

With some of the more complex builds, there is still too much bounce. The cat from before never stays on the ground.

https://github.com/tsunamayo/Starship-EVO/assets/10069469/48e3b41c-5e9d-4025-9237-2d8404668613

**Comment by Alkaliii** [comment]
Good Improvements all around. Finally got one to work.
[microwalker2.zip](https://github.com/user-attachments/files/15845010/microwalker2.zip)
https://imgur.com/Hw0fre3

Is a keyframe at 0.5 required, I seem to get some buggy looping when I don't have one.

**Comment by asanagisae** [comment]
A few players had some luck with smaller 2-legged walkers, so I thought I'd try one too.

It's moderately stable, but it keeps bouncing on the floor when standing still. Perhaps it would be better if the "adjust foot to match terrain slope" feature could have a slider, so we can reduce the effect or turn it off?

https://github.com/tsunamayo/Starship-EVO/assets/10069469/28e78af1-3194-4457-b50b-248d9b0732ea

**Comment by tsunamayo** [comment]
@asanagisae Hi, nice looking mech! Send me the bp and I will investigate and fix the issue.

**Comment by asanagisae** [comment]
_<does a dance>_

https://github.com/tsunamayo/Starship-EVO/assets/10069469/7bb3cf10-2a39-4cf1-a7e3-90d0a7e21b9d

Here you go! The current BP, while modified, still bounces when not walking.

[Mechawalker.zip](https://github.com/user-attachments/files/15857907/Mechawalker.zip)

I tried to tweak the walk cycle so that the speed remains positive for the whole time. It seems to always take the lowest point on the mech, assuming that at least one foot will always be touching the floor. Faster gaits, such as a horse's gallop, require the creature to be off the ground entirely for part of the walk cycle while the legs move forward in position.

When I set the walk speed to 0 (zero assisted forward momentum), it seems the mech only moves sideways when pressing W, if at all. Does this graph have any meaning? If we optimize the walk cycle for a continuous positive speed, do we get a speed boost?

![image](https://github.com/tsunamayo/Starship-EVO/assets/10069469/d3a915b6-784f-4863-b7bf-34aeb9911cbe)

**Comment by tsunamayo** [comment]
@asanagisae yeah I am not sure how you made it jump, the terrain preview wont deal with that for sure. You need to imput a bit of speed otherwise it wont work.

**Comment by tsunamayo** [comment]
also @asanagisae send me the bouncing cat, when you show me something to debug always send me the faulty blueprint.

**Comment by asanagisae** [comment]
@tsunamayo Apologies. It's the same cat from before. I thought you had it from the previous issue. I've been afraid to put more work into it in case it is fundamentally incompatible with the animation engine design.

[Mechaquad.zip](https://github.com/user-attachments/files/15860445/Mechaquad.zip)

I think the 2-legged walker's main issue is the inability to stand still. Is that something that can be mitigated, taking into consideration the mech designs from other players? Should we have more players submit mech designs to accommodate a broader range of mech feet?

**Comment by tsunamayo** [comment]
@Alkaliii yep you need a key frame at 0.5 if you use symmetry, so that the symmetrical key starts at 0. I will see what I can do to enforce that.

**Comment by tsunamayo** [comment]
@asanagisae Yeah sorry np it s hard to keep track of everything on my side! I am having a look at it now, I keep you posted.

**Comment by tsunamayo** [comment]
@asanagisae So right of the bat the config is done properly, the foot are set to hip, and the too knee are set at knee instead of something like knee and custom / knee 1 for example.
That being set the foot placement are extremely twitchy when not used in mech, so I am looking into that too. Cheers

**Comment by asanagisae** [comment]
Thanks for looking at it. Yes, the joint ordering may be a little strange. 😄 It was designed and built in phase 1 when custom joints weren't possible. Joints with the same type are meant to follow the same animations and key frames for this style of walk to be possible. The types aren't indicative of their purpose - hips as feet, for example.

That said, the actual joint names shouldn't affect the mech's behavior? Does the engine require legs to be set up as hip > knee > foot?

**Comment by tsunamayo** [comment]
@asanagisae Yep it does. Foot have the foot terrain alignment feature for example, hip have sound effet (you can toggle it on with a custom join though).
I fixed the twitchy issue, it is was due to the moment of inertia being too small as being on the main parent entity only. I still need to compute an accurate center of mass though.
